esphome/components/graph/__init__.py

```diff
@@ -0,0 +1,107 @@
+import esphome.codegen as cg
+from esphome.cpp_types import App
+from esphome.const import (
+    CONF_ID,
+    CONF_WIDTH,
+    CONF_SENSOR,
+    CONF_HEIGHT,
+    CONF_MIN_VALUE,
+    CONF_MAX_VALUE,
+    CONF_MIN_RANGE,
+    CONF_MAX_RANGE,
+    CONF_LINE_THICKNESS,
+    CONF_LINE_TYPE,
+    CONF_UPDATE_INTERVAL,
+    CONF_X_GRID,
+    CONF_Y_GRID,
+    CONF_BORDER,
+    CONF_AXES,
+)
+
+CODEOWNERS = ["@synco"]
+
+DEPENDENCIES = ["display"]
+MULTI_CONF = True
+
+LineType = display.display_ns.enum("LineType")
+LINE_TYPE = {
+    "SOLID": LineType.LINE_TYPE_SOLID,
+    "DOTTED": LineType.LINE_TYPE_DOTTED,
+    "DASHED": LineType.LINE_TYPE_DASHED,
+}
+
+Graph_ = display.display_ns.class_("Graph")
+graph_ns = cg.esphome_ns.namespace("graph")
+GraphAxes = graph_ns.class_("GraphAxes")
+GraphAxesPtr = GraphAxes.operator("ptr")
+
+GRAPH_BASIC_SCHEMA = cv.Schema(
+    {
+        cv.Required(CONF_ID): cv.declare_id(Graph_),
+        cv.Required(CONF_WIDTH): cv.positive_int,
+        cv.Required(CONF_HEIGHT): cv.positive_int,
+    }
+).extend(cv.polling_component_schema("10s"))
+
+CONFIG_SCHEMA = GRAPH_BASIC_SCHEMA.extend(
+    {
+        cv.Optional(CONF_SENSOR): cv.use_id(sensor.Sensor),
+        cv.Optional(CONF_MIN_VALUE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MAX_VALUE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MIN_RANGE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_MAX_RANGE, default=float("nan")): cv.float_,
+        cv.Optional(CONF_LINE_THICKNESS, default=3): cv.positive_int,
+        cv.Optional(CONF_LINE_TYPE, default="SOLID"): cv.enum(LINE_TYPE, upper=True),
+        cv.Optional(CONF_X_GRID, default=float("nan")): cv.float_,
+        cv.Optional(CONF_Y_GRID, default=float("nan")): cv.float_,
+        cv.Optional(CONF_BORDER, default=True): cv.boolean,
+        cv.Optional(CONF_AXES): cv.All(
+            cv.ensure_list(
+                {
+                    cv.GenerateID(): cv.declare_id(GraphAxes),
+                    cv.Required(CONF_SENSOR): cv.use_id(sensor.Sensor),
+                    cv.Optional(CONF_MIN_VALUE, default=float("nan")): cv.float_,
+                    cv.Optional(CONF_MAX_VALUE, default=float("nan")): cv.float_,
+                    cv.Optional(CONF_MIN_RANGE, default=float("nan")): cv.float_,
+                    cv.Optional(CONF_MAX_RANGE, default=float("nan")): cv.float_,
+                    cv.Optional(CONF_LINE_THICKNESS, default=3): cv.positive_int,
+                    cv.Optional(CONF_LINE_TYPE, default="SOLID"): cv.enum(
+                        LINE_TYPE, upper=True
+                    ),
+                }
+            ),
+            cv.Length(min=1),
+        ),
+    }
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID], config[CONF_WIDTH], config[CONF_HEIGHT])
+
+    sens = await cg.get_variable(config[CONF_SENSOR])
+    cg.add(var.set_sensor(sens))
+    cg.add(var.set_min_value(config[CONF_MIN_VALUE]))
```

Comment:
(2/3) Like this, but for all optional fields with no defaults

```suggestion
    if CONF_MIN_VALUE in config:
        cg.add(var.set_min_value(config[CONF_MIN_VALUE]))
```